### ISSUE TITLE
Replace Glide64 with GLideN64

### ISSUE THREAD

**Issue by Fanatic-64** [project64/project64#428]
The public version of GLideN64 has been released, along with source code: https://github.com/gonetz/GLideN64/releases/tag/Public_Release_!

It should replace Glide64 For PJ64, per zilmar's plans, and because it's overall better.

This issue can be used to track that task.

**Comment by Lithium64** [comment]
Great plugin but won't work here

**Comment by AmbientMalice** [comment]
Unfortunately, the plugin isn't mature enough to replace Glide64 just yet. Frankly, it's not mature enough to incorporate into PJ64 yet. LLE is broken badly, and heaps of games have major issues.

Two FB effects are broken/missing in Mario Tennis, which is kinda bad considering MT was one of the plugin's selling points.

**Comment by Lithium64** [comment]
@AmbientMalice These issues ain't fixed on public release?

**Comment by AmbientMalice** [comment]
@Lithium64 Nope. I tested Mario Tennis last night and the VS screen is missing the upper half still - this bug was noted on the github Issues a while ago.

Also, LLE has some nasty problems. Some inherited from Ziggy's, such as lines used for map rendering in Turok/Doom/Winback/etc being garbled -- others original - Indiana Jones has no particle effects/2D sprites rendered, for example.

I'm hoping that there are some skilled coders who can plug these gaps because Gonetz bit off more than he could chew, IMO.

**Comment by Fanatic-64** [comment]
So I should forget about it and close this?

**Comment by AmbientMalice** [comment]
>So I should forget about it and close this?

No, IMO. GLideN64 should eventually become PJ64's primary - possibly only video plugin. It just needs some TLC before it's acceptable.

**Comment by LegendOfDragoon** [comment]
>No, IMO. GLideN64 should eventually become PJ64's primary - possibly only video plugin. It just needs some TLC before it's acceptable.

I can understand it becoming the primary video plugin, but I don't think it should be the only video plugin, because not everyone is able to use that plugin, unfortunately.

**Comment by Fanatic-64** [comment]
I think Jabo should be the primary plugin, and GLideN64 should be available for those with modern PCs.

**Comment by AmbientMalice** [comment]
>I think Jabo should be the primary plugin, and GLideN64 should be available for those with modern PCs.

Jabo is riddled with problems and missing features, and it is also closed source, making it a development dead end. TBH, using Jabo as the default instead of Glide64 is already causing problems. It's not ideal to have an endless stream of people complaining that <insert game here> doesn't render correctly. Worse, people who unwittingly use Jabo without realising just how lacking it is. (Aka, the average person who just downloads PJ64.)

Not to say Jabo isn't a venerable plugin that compensates for its poor accuracy and compatibility by being fast and widely supported--but it is a very poor choice of default plugin in my opinion.

**Comment by Lithium64** [comment]
In my opinion should be

Glide64 primary plugin
GlideN64 for modern PCs

**Comment by AmbientMalice** [comment]
@Lithium64 TBH, a lot of the current performance discussion is a bit fruitless simply because GLideN64 performs so badly that a "modern PC" is going to struggle with it. In an ideal world, GLideN64's awful performance would be fixed and a simple check put in place to make Glide64 the default plugin if the user's GPU doesn't support GL 3.3+

GLideN64 performs shockingly poorly. Until someone fixes the performance, the plugin is practically useless for general use. In fact, the performance problems are far more serious than game compatibility issues.

**Comment by Predator82Germany** [comment]
GLideN64 is not so bad

**Comment by AmbientMalice** [comment]
@Predator82Germany

The current performance issues are atrocious. Vigilante 8 running at ~40VIs on a 3.3Ghz CPU? In HLE? In some cases, GLideN64 runs games with worse performance in HLE than Jabo's ran them in LLE. Not to mention all the other games afflicted by unplayable performance the moment frame/depth buffer effects are involved.

Make no mistake - I hold GLideN64 in high regard, and expect good things from it in the future. But currently, it is in poor shape performance-wise. I'm very sceptical about how well Android devices will handle it - unless the underlying issues are fixed.

**Comment by Talantyyr** [comment]
Just to mention it, GlideN64 does rely on your GPU, not CPU, so you need to have a fast graphics card for it.

**Comment by AmbientMalice** [comment]
@Talantyyr It relies on both, especially for framebuffer effects.

However, the real root of GLideN64's performance problems is actually Nvidia's adaptive power management. Setting to Maximum Performance cures the awful stuttering and problems with higher resolutions.

**Comment by Talantyyr** [comment]
Oh ok, gonetz keeps saying that cpu is not that important. On my old i5 with a AMD R9 290 it runs very well, but on my surface pro 3 (core i3 / Intel HD4000) it's stuttering, though i have great perfomance with other plugins. That's why i thought he was right.

**Comment by CraigClements** [comment]
GLideN64 actually works amazingly well for those with bleeding edge hardware, far better than Glide64 overall. Games that GLideN64 struggles with, Glide64 similarly struggles, bar a handful.

Nonetheless, I do agree that Jabo be the main plugin right now, Glide64 should be the main plugin for at least the near future, and GLideN64 / some LLE plugin in the further future when PC hardware catches up to the requirements. A lot of people who are using PJ64 are not also hardware enthusiasts and are running on laptops and such, so making GLideN64 the default right now will lead to forums everywhere getting messages that they are experiencing lag.

**Comment by project64** [comment]
it would be nice to do some type of detection of the system and try to tailor default to the system, so it could be gliden64 on high power machines, default full screen res same as desktop res, etc

**Comment by AmbientMalice** [comment]
>and GLideN64 / some LLE plugin in the further future when PC hardware catches up to the requirements.

GLideN64 *is* an LLE plugin, though. And at what point is the cutoff for establishing hardware has caught up? GLideN64 runs on 6 year old CPUs and 5 year old GPUs with no issue. Maybe even older hardware. (Since it runs on GL 3.3 cards, too.)

You're always going to have people with crummy PCs, which is why keeping Glide64 around for legacy support is a good idea. But I don't think it's wise to excessively cripple PJ64's game compatibility for the sake of clueless people. (Jabo is the reason PJ64 has such a bad reputation in terms of bad graphics emulation.) I'm not sure how the power management problem with GLideN64 should be solved. But considering Dolphin is also affected, it's not THAT serious an issue.

GLideN64 has problems. But these can in theory be fixed. Glide64 is likely always going to be afflicted with its corrupt texture nonsense and wonky framebuffer support.

>it would be nice to do some type of detection of the system and try to tailor default to the system, so it could be gliden64 on high power machines, default full screen res same as desktop res, etc

You could just offer a shiny prompt as part of a Wizard when you first boot PJ64. Defaulting to desktop resolution is a VERY good idea, though.

I'd actually like to see an automated hack that enables widescreen mode for games with native widescreen when running in widescreen resolutions.

**Comment by CraigClements** [comment]
I have an AMD A8 APU laptop and would use Jabo's D3D8 over Glide64 on it almost every time as it is much faster. It is not old hardware (bought within the last 2 years), and that is the kind of hardware people are using. Not every one also has a dedicated gaming rig or uses one for PJ64.

**Comment by AmbientMalice** [comment]
@CraigClements Is Glide64 slow on your PC? That's a bit odd. Glide64 was originally designed for Voodoo GPUs. It has some CPU overheads, particularly for software depth buffering (optional, but turned on by default to prevent depth bugs), but the GPU load is minimal.

The primary reason Jabo is fast and runs at 500-800fps on my PC is because it is lacking crucial features. Proper depth calculation, for example. Or framebuffer support. (It can do *some* FB effects, but anything per-frame is useless.)

I'm a big believer in backwards compatibility, but not at the cost of making an extremely broken, closed source video plugin the default. People should not be using Jabo if they can avoid it.

**Comment by LegendOfDragoon** [comment]
I have to agree with @CraigClements . Jabo's performance is much better, so I use that way more often than Glide64. I also get less crashes with Jabo's plugin and I get a black screen when switching games with Glide64 ;/ .

**Comment by AmbientMalice** [comment]
@LegendOfDragoon You're not bothered by the fact Jabo runs almost no games correctly? It can't even emulate Mario 64 correctly.

Mind you, what are your PC specs?

**Comment by LegendOfDragoon** [comment]
@AmbientMalice I am bothered by the lack of accuracy, but that's the price I have to pay for performance. Fortunately many of the games I play, look decent enough for my standards with Jabo's.

For certain games like Quest64, Star Wars Ep1 Racer, RE2, I'd use Glide64 because the difference in accuracy is much more noticeable to me, and the performance isn't bad either, compared to other games I play like Super Smash Bros. 

My PC specs are weak ;/ . 2.67 Ghz quad core i5, Intel integrated gfx, 8gb RAM, and I only have OpenGL 2.1 lol.

**Comment by AmbientMalice** [comment]
>My PC specs are weak ;/ . 2.67 Ghz quad core i5, Intel integrated gfx, 8gb RAM, and I only have OpenGL 2.1 lol.
The only real problem with your specs is the Intel GFX. An Intel i5 is more than enough for most purposes, even at 2.67Ghz.

**Comment by cxd4** [comment]
Usually I use Jabo's plugins unless I'm using something else, like Azimer's Audio or pixel-accurate.

Tend to be using PJGlide64 more often now that it's a part of the package...usually high-res and bug-free enough to my liking to play some games in hi-res every once in a while.  I keep forgetting to try GlideN.

**Comment by CraigClements** [comment]
Glide64 can be slow on some of my 'weaker' systems like my primary laptop (the AMD A8), yes, mostly due to the wrapper overhead (the default one, not a 3rd party, never tried that personally). Glide64 is pretty fast using something like a Voodoo 5 with no wrapper as far as I know. I haven't tried GLide64 on all of my systems, but I don't suspect it would run very well on almost all of them barring my fastest system that was built around emulation/gaming.

The speed is game dependent, as with some games, the A8 can still cope Glide64. I generally take 60 V/Is for the vast majority of the time to be considered good enough to play. Also depends on if I am running off battery, but I rarely do that nowadays. When testing on my netbook (Intel Atom), I would fall all the way back to Jabo's D3D6, but obviously I would not recommend something like that be default in this day and age.

I absolutely do agree with the philosophy to move to open source plugins rather than sticking with Jabo's existing plugins forever - I think that is really important. Obviously the real question is how soon do we move on to these new, improved open source plugins - which is a bit more of a matter of opinion. Glide64 is a great choice to consider as a replacement to Jabo's. Its a bit difficult to gauge as pretty much anyone who posts on a forum for help often have lower specs, so it skews it a bit.

**Comment by Lithium64** [comment]
We are in 2015, keep a plugin made for hardware manufactured ten years ago or even older as default will never be a good idea.

We don't need to support legacy systems like Windows XP and outdated APIs, few people still use these systems which are no longer supported, anyway older versions of pj64 and plugins are always available for those who don't want or can't update your system and hardware.

**Comment by oddMLan** [comment]
I think this should be closed as wontfix, or at least until GLiden64 can be modified to support older devices with lower OGL extensions support with reduced graphic accuracy.
Or bundle GLideN64 as-is currently as an alternate, higher-end option for users so they don't have to do the run-around to get it... if they are even aware of its existance in the first place.

But... #536 
https://github.com/project64/project64/issues/652#issuecomment-145072521

**Comment by cxd4** [comment]
If it were my repo probably I would close this, but it's probably at least as smart of a move that project64 does not close this or else it would just spark fighting over Glide vs. GLideN or whatever due to the issue being closed when some people don't want to see it closed.

Regardless, I agree since a) it's "wontfix" like you said b) it's not even an issue; it's a feature request.

**Comment by project64** [comment]
yes it will ventually be a modified version of glide64 instead of glidn64

**Comment by AmbientMalice** [comment]
>yes it will ventually be a modified version of glide64 instead of glidn64

That seems like a bad idea since it is highly unlikely Glide64 will ever be anywhere near as accurate as GLideN64. Glide64 is a dead end. There's basically no reason to use it at this point except to run the handful of games GLideN64 struggles with. It performs worse, and its accuracy is horrific.

**Comment by project64** [comment]
well the version of glide64 will be my own fork .. where I will clean up the code and get things working the way I want and more intergrated with pj64, especially for android. Like I can do direct game hacks that is only meant to work for pj64. 

While GlideN64 is really good I dislike it using so much cutting edge open gl features, prefer greater compatibility and using the lowest possible specs

**Comment by AmbientMalice** [comment]
>While GlideN64 is really good I dislike it using so much cutting edge open gl features, prefer greater compatibility and using the lowest possible specs.

GLideN64 is designed to be as backwards compatible as possible, though. It uses higher end GL features in order to emulate things that are impossible to emulate otherwise.

**Comment by project64** [comment]
the aim is to do everything as accurately as possible per game hacks are a last resort .. but things like mario kart, the CPU copies a buffer and with out some type of information about that (settings/hack) then there is no way the GFX plugin can do it. Unless it writing all the data to the RDRAM instead of the video card memory.

**Comment by AmbientMalice** [comment]
>Unless it writing all the data to the RDRAM instead of the video card memory.

Doing that with async copies really isn't that demanding. At least for desktop. Android is another matter, but you're already running into problems such as lower end Android devices struggling with accurate blend emulation.

**Comment by Talantyyr** [comment]
Zilmar, have you considered putting the glideN64 plugin in the android port as well? (with selectable plugins in the android app or maybe even via autodetection of the GLES capabilities?)

It's completley understandable that you want to support old mobile phones as well, but i think most people will use rather powerful android devices to play N64 games like the Nivida Shield Tablet or TV.

**Comment by LegendOfDragoon** [comment]
>GLideN64 is designed to be as backwards compatible as possible, though.

It currently just crashes on my laptop. Can't use it at all anymore, on that.

For what zilmar wants to do, it makes sense to go with Glide64. I'm hoping that porting Angle will be a success. I'm not a big fan of OpenGL and I personally would rather not use Glide64 or GLideN64. I'm actually liking Rice's Video plugin and am having fun with that. If some of Glide64's issues after Angle is ported, I will gladly devote time to actively contributing to it.

**Comment by AmbientMalice** [comment]
>It currently just crashes on my laptop. Can't use it at all anymore, on that.

It's in a state of flux, partially because of a lot of changes made to improve GLES support. One of the bigger issues is the plugin use GL compatibility profile instead of GL core profile. That should be fixed in the near future.

>I'm not a big fan of OpenGL and I personally would rather not use Glide64 or GLideN64.

GL is mandatory for the plugin to operate outside of Windows. Also, porting some of the extremely accurate features like accurate N64 depth to DX could prove tricky.

**Comment by LegendOfDragoon** [comment]
>It's in a state of flux, partially because of a lot of changes made to improve GLES support. One of the bigger issues is the plugin use GL compatibility profile instead of GL core profile. That should be fixed in the near future.

Considering the fact that it has never supported older hardware as good as other OpenGL plugins, I doubt it will ever properly support anything less than OpenGL 3.x. The fact that I'd rather use older builds is concerning. I tested Mario Kart on PJ64 and had the same performance issue when using sync mode, so I'm fairly certain there's some regression in the code.

>GL is mandatory for the plugin to operate outside of Windows. Also, porting some of the extremely accurate features like accurate N64 depth to DX could prove tricky.

I'm not a big fan of Mupen, so I feel that there would not be much of a downside in being Windows only. I've dealt with too many issues in OpenGL, so I feel that it's not worth my time to code in OpenGL. I know it won't exactly be easy to code in D3D, but I think it's worth a try.

**Comment by project64** [comment]
The version I am working on has to support android, I do not think android has anything other then OpenGL ES ...

So it will be open gls es, with a wrapper for windows (angle). Granted the wrapper can do d3d or opengl

**Comment by LegendOfDragoon** [comment]
How well does GLideN64 perform on Android in comparison with Glide64? I was under the impression that GLideN64 might be more suitable than Glide64 for android, but I could be wrong.

>The version I am working on has to support android, I do not think android has anything other then OpenGL ES ...
So it will be open gls es, with a wrapper for windows (angle). Granted the wrapper can do d3d or opengl

I understand that OpenGL ES is the only feasible option for Android. There's Vulkan, but that doesn't support older devices and older versions of Android OS. I just happen to not be that interested in Android, so I'd personally prefer just using D3D without using any wrappers. For what you're trying to do, I think you're making the right decision. I really hope Angle works well.

With Glide64, I feel like I'd end up spending too much time cleaning it up, if I were to use that as a base. But if others are going to also working on it, I don't mind using it as a base. Basically if I was doing everything on my own, I believe I'd be better off writing my own plugin. Having a team, makes a big difference though.

One issue I see with Glide64 is that it has performance issues that aren't due to the wrapper. For instance in Bangaioh, I get better performance with Angrylions (without VI filters) than I do with Glide64. That's a big deal considering Angrylion's code has very bad performance for 2d games. I profiled and found the code responsible for the bottleneck. I could easily just write an optimized version of the function in SSSE3, but I feel that there are better approaches, especially since it's a hardware renderer.

**Comment by project64** [comment]
I know I am going to spend a lot of time just cleaing up the code, now I got angle working somewhat

**Comment by Frank-74** [comment]
Just tried to build the angle branch, I get 100's of identifier not found errors, after adding the angle include path.

**Comment by cxd4** [comment]
>I'm not a big fan of Mupen, so I feel that there would not be much of a downside in being Windows only.

Who said anything about Mupen64 in this entire thread before you did?
Granted whatever that has to do with anything.

Project64 has already been running on Linux on the sole merit that Android really is nothing better than Google's redistribution of the Linux kernel.  Android in that respect is in fact just another Linux "distro".

The same will be true for the other distributions of the kernel, Linux, that Project64 will in the future run on.  It's just not that way now because zilmar spends all his priority time away from that in favor of the ARM dynarec, ANGLE etc..  If he gets bored and lifeless enough he'll eventually research enough to reject or accept my previous pull request so I can move on with what to do next.  Not that that will happen but still. :)

**Comment by cxd4** [comment]
Frank, "identifier not found" errors can refer to different things.  Sometimes compile errors but IIRC (could be wrong on this since I haven't used MSVC in some months) linker errors with the libraries you're using.  Even if they are compile errors there are still multiple things that can cause them.

I would copy-paste the entire build log with all the errors in it to upload the text to gist.github.com, pastebin or whatever you prefer and link that here.  That would give a better idea what's missing.

**Comment by Frank-74** [comment]
@cxd4  I'm missing the DX11 libraries angle needs, which comes with the Windows 7, 8, 8.1 or 10 SDK.

**Comment by LegendOfDragoon** [comment]
>Who said anything about Mupen64 in this entire thread before you did?
Granted whatever that has to do with anything.

I deem Mupen as currently the most usable emulator on Linux and basically was saying that there's no emulators on Linux that I'd want to use.

@Frank-74 you might need MSVC 2015.

**Comment by project64** [comment]
I think I added them all to the repository so you should not need the DX libraries installed .. but ANGLE is VS 2015 only as @LegendOfDragoon  suggested.

**Comment by Frank-74** [comment]
Installed VS2015 and compiled the angle branch.  I get RunRSP errors when closing config.  Also black screen when switching to fullscreen, RunRSP error on exit fullscreen.

Windowed works well, but have to restart Project64 to play another game.  Few funky problems, models/characters are missing in Resident Evil 2, and first half of Perfect Dark intro is upside down:
![glide64_perfect_dark_02](https://cloud.githubusercontent.com/assets/11056843/21753087/affc80ec-d5dc-11e6-9abf-cc33fb58b278.jpg)
Early days, but performance is good.

**Comment by LegendOfDragoon** [comment]
@Frank-74 what OS, cpu, and gpu are you using? It doesn't work for me unless I boot it from Visual Studios for some reason.

>Early days, but performance is good.

Is the performance is better than w/o Angle?

**Comment by Frank-74** [comment]
Laptop, Windows 8.1, Quad core Pentium N3540 @ 2.16Ghz, built in Intel HD Graphics.

Performance of Project64 Video (as it's now called), seems smoother, steady 60VI's throughout the Perfect Dark intro.  It looks a bit more shiny too, reflections that is.  I haven't tested that much, cannot go fullscreen either yet.

**Comment by LegendOfDragoon** [comment]
I tried testing on my desktop and it still had the same issue. I wonder if it's the compiler settings I used. I checked the dll dependancy and I see d3d9.dll and not d3d11. I wonder if I'm simply running the d3d9 code and the d3d11 code works fine. Of course the issue needs to be fixed regardless of whether the d3d11 one works.

**Comment by project64** [comment]
@Frank-74 thanks for testing and looking at this .. I am starting to look at the resolution code which will also should fix full screen when done. I will look at the crashes when stopping after that. 

There is a good reason this is staying in a branch for a while :)

Yes it should have all the Android GLES version bugs as it is using the android version of glitch, I wonder if perfect dark upside down is in android as well. 

At least I feel Like I can start changing the code base of glide64 to clean it up, with all these changes I renamed it. (also in part I can switch between branches and compile glide64 original one and the new angle version one).

the other things angle gives is the possibility to use d3d9, d3d11, glide and vulkan (still that is in early days). I need to add a config option to select which render to use.

**Comment by Frank-74** [comment]
When the motion blur effect happens in the Perfect Dark intro, the blur is the right way up, main video still upside down, but then the next scene video is the right way up.  Also, I don't think fog was working when I tested Monaco Grand Prix.  Monaco Grand Prix and Goldeneye have worse performance than PJGlide, Perfect Dark seems better performance.  No tearing though, that's a bonus.

**Comment by Frank-74** [comment]
Perfect Dark intro is not upside down in Android, just tested.

**Comment by project64** [comment]
Weird with them being different, seems like maybe there is some bug in the code that just does not effect android. 

I know I will spending a lot of time cleaning up the code and then trying to improve/getting stable. But still will be a while I think before angle is default. After the next release I will start changing the main line more and the branch as well.

**Comment by cxd4** [comment]
>I deem Mupen as currently the most usable emulator on Linux and basically was saying that there's no emulators on Linux that I'd want to use.

...in citation of a reason why _"there would not be much of a downside in being Windows only"_.

So the reason why code being Windows-only wouldn't have much of a downside is because the emulators you like which are available on Windows haven't finished being ported to Linux yet?

**Comment by cxd4** [comment]
>the other things angle gives is the possibility to use d3d9, d3d11, glide and vulkan (still that is in early days). I need to add a config option to select which render to use.

Wow, that's pretty cool.  I didn't guess ANGLE had an option to wrap on top of Vulkan and GLIDE.

**Comment by project64** [comment]
was meant to be opengl not glide .. but does have some vulkan support.

Should be able to work for linux as well to be able to use opengl

**Comment by cxd4** [comment]
Oh.  Well Vulkan is amazing, but I was going to say, GLIDE would have been funnily amazing too.

I didn't guess OpenGL itself would be the other one since GL ES is almost a complete subset of that, so I don't imagine a very indirect routine to "wrapping" over that.

Anyway, we have several graphics plugins that run already on Linux, before this one.  That Windows-specific stuff just needs to go before the Linux port of Project64 runs.  Then again I think all of it is gone, but sometimes replaced with Android-specific stuff, even though Android really is just another Linux.

**Comment by Frank-74** [comment]
Hardware frame buffer causes the PD intro to be upside down.  Turn it off and it's the right way up, but 4-5 VI's instead of 60.

**Comment by LegendOfDragoon** [comment]
I did more testing with GLideN64 yesterday and the performance is a big concern. It's honestly concerning that z64gl is faster in some cases.

@Frank-74 you think you could upload an Angle build? I'm kind of puzzled why it seems to work better on your end. I'm wondering if our builds are different, somehow.

>So the reason why code being Windows-only wouldn't have much of a downside is because the emulators you like which are available on Windows haven't finished being ported to Linux yet?

For me it's more like the nail in the coffin. I've just dealt with way too many setbacks, for me to really care about supporting non-Windows OSes. Even last night, I made another unsuccessful attempt to compile plugins on Linux.

I don't believe PJ64 will be fully ported to Linux anytime soon and I'm not willing to just faithfully wait either. I'm pretty sure exception handling is different on Linux. Since I'm likely not going to get much help from others, I might as well code something that will be the most beneficial for myself to use.

**Comment by Frank-74** [comment]
@LegendOfDragoon 
Sure, here's my build of the angle branch Glide.
https://www.dropbox.com/s/0k5ezvh84jholhf/Project64Video.dll?dl=1

I never touched any compiler settings.  Just opened the sln, selected Release and build solution.

**Comment by project64** [comment]
> I don't believe PJ64 will be fully ported to Linux anytime soon and I'm not willing to just faithfully wait either. I'm pretty sure exception handling is different on Linux. Since I'm likely not going to get much help from others, I might as well code something that will be the most beneficial for myself to use.

Exception handling is different it uses signals .. but the core handles that all fine now .. it is like 98% done for linux, but as I do not have linux I do  not care that much about finishing it for it. Really just needs a UI to link it all together, an audio and input plugin. The audio plugin might be very similar to the android one. So maybe just UI and input.

**Comment by cxd4** [comment]
>I'm pretty sure exception handling is different on Linux.

And surprise, Android is Linux like I said.  I'd be pretty surprised if exception-handling on Linux wasn't the same as Android.  lol because that n00b torvalds copy pasta'd off of google's operating system.

>I've just dealt with way too many setbacks, for me to really care about supporting non-Windows OSes.

That's fine, but that means that for you there wouldn't be a downside to being Linux-only, not that there wouldn't be a downside to being Linux-only for Project64 in general.

I had many setbacks learning Linux.  My family raised me on DOS and Windows 16-bit.  The challenge involved with switching over was only a setback in the short-term sense; it's quite logical to assume that something that starts out as a pain, with modern operating systems, is not at all supposed to stay a pain, or there is something wrong with my brain if I think otherwise, because it's so well-designed for everyone else except me.

>Even last night, I made another unsuccessful attempt to compile plugins on Linux.

That depends what you mean.  I would be more than happy to assist you with that.

Depending on what you mean I'd like to hear more.  Compiling on Linux is only more out-of-the-box faster than compiling on Windows for the reason that there are no compilers shipped with Windows (you have to install everything)...but that goes for the libraries and headers you need to compile in certain APIs when building certain plugins on either operating system regardless.

**Comment by LegendOfDragoon** [comment]
>That depends what you mean. I would be more than happy to assist you with that.

I basically want to start testing z64gl on Linux. In the past, I had all sorts of trouble compiling it and gave up on that. Last night, I tried compiling your RSP and it was giving me errors. I assume make.sh should allow it to compile on Linux out of the box right?

I wish I could just download binaries lol.. Anyway I would first like to be able to compile your RSP and then hopefully z64gl as well, so I can finally start testing that.

>That's fine, but that means that for you there wouldn't be a downside to being Linux-only, not that there wouldn't be a downside to being Linux-only for Project64 in general.

I didn't realize it was that close to being fully functional. I figured there was more work to be done than just UI and input & audio plugins. Once a basic UI is done, even I'll somewhat care about Linux support now :smile: . Who knows, maybe i'll get the rsp recompiler working on Linux.

**Comment by cxd4** [comment]
>Who knows, maybe i'll get the rsp recompiler working on Linux.

That would be good since then I could get faster LLE without using this interpeapyearyerstrhsthdg one plugin that just hasn't been optimized to be as fast as recompilers yet.

>I wish I could just download binaries lol..

https://en.wikipedia.org/wiki/Unix_philosophy

I seem to forget which number rule this is, but in any UNIX-based operating system (which basically means everything except Windows), you should try to prefer text-based streams.  Everything becomes just so much more extensible and portable.  There's even a Linux out there called Gentoo, and others, that don't have binary-based software packages.  It installs packages by you invoking the install command, but what it's really doing is compiling everything.  Binary format is cool and efficient, yes, but there's an acquired taste to be had for both.

>I basically want to start testing z64gl on Linux. In the past, I had all sorts of trouble compiling it and gave up on that.

All I've ever done is `cd` into the directory that contains the file called `Makefile` (you should always do this for projects built from Makefiles), and execute `make` on the command line.

If that's not good enough please do file an issue on purple's repo (instead of being a fallaha56).

>Last night, I tried compiling your RSP and it was giving me errors. I assume make.sh should allow it to compile on Linux out of the box right?

Yes, but remember that for OS and workstation security, stability and simplicity, typing executable names on the command-line in Linux (and most other UNIX-derivatives) _does not search in the current working directory_.  It's not like Windows, where Windows searches system32/wow64, then system, then %userprofile%, then current working directory or however it works I forgot.

So if you typed ONLY `make.sh` then it looked for a file named that in root directories on your drive like `/bin/`, `/sbin/`, `/usr/lib/`, etc., _not your CD you set the command prompt to_.  (`make` for z64gl is different since that's a system utility installed to your root dirs.)

So you're supposed to enter ./make.sh not make.sh.  Windows and DOS also used and copied the `.` notation for referring to the current directory, just like `..` means parent directory.  The difference is on Windows a) it would be `..\make.sh`, not `../make.sh` with the forward slash, b) it's optional on Windows.  If you want to know why Windows has it optional, Google it, I don't know.

**Comment by LegendOfDragoon** [comment]
>(instead of being a fallaha56)

Not sure what you mean by that lol, but the last time i tried compiling z64gl was before purple even created a repo. I just tried compiling and it compiled out of the box. I just figured I shouldn't bother trying to compile z64gl again until I compiled rsp first. I'm glad purple made it easy to compile :smile: .

>So you're supposed to enter ./make.sh not make.sh.

Ah I see. Got it compiling now. I was used to just using make. Didn't know I had to do ./ .

Not sure why I get black screen. Tried booting a few different games and it eventually froze my computer... I'll try again when PJ64 has a UI in Linux :smile: .

**Comment by project64** [comment]
that is also assuming hle audio for linux as the RSP plugin for windows needs a good refactoring and needs a lot of work if it was to run on ARM

**Comment by LegendOfDragoon** [comment]
Now that I remember testing Glide64 in Mupen on Linux, I doubt I'll see a big difference in z64gl's performance on Linux..

>that is also assuming hle audio for linux as the RSP plugin for windows needs a good refactoring and needs a lot of work if it was to run on ARM

I know Ziggy forked PJ64 1.4's RSP and got that compiling on Linux. Looks to me that it wouldn't require too much work. Then again PJ64 2.x's RSP has a different UI, so maybe there's a little more work required than I remember.

**Comment by project64** [comment]
x86/x86-64 linux should not be much work .. mostly bug fixes since then ... 

but if it was to run on any other platform then it will need more work

**Comment by cxd4** [comment]
>Not sure why I get black screen. Tried booting a few different games and it eventually froze my computer...

That sounds like you're using the recompiler.  I thought I mentioned last year when you were getting it to work on Linux back then, that you cannot use that.

Somewhere on EmuTalk Hacktarux posted the open-source patch to get the dynarec to work again on newer distros of Linux, but I never recorded it since I just test the interpreter.

>>(instead of being a fallaha56)

>Not sure what you mean by that lol,

Aw, come on.  You don't suspect I was comparing you to him, do you?

All I was meaning was report the issue you have with compiling and try to help me make things easier if there's some special problem you have compiling it, don't tl;dr logic and say LOL SEE WHY DONT YOU MAKE THINGS EASIER.  Because that really makes the troubleshooting process a lot more distracting and helps prevent getting things resolved.

Saying that I need to host a Dolphin-like build-bot because compiling works on X and Y but not Z is like saying you need to compile for Z because the Mac binary doesn't work on my Windows.

>Now that I remember testing Glide64 in Mupen on Linux, I doubt I'll see a big difference in z64gl's performance on Linux..

It doesn't matter.  As far as video plugin bottlenecks go I don't think you should expect a difference when switching between OSes, at any rate.  Certainly not on an Intel video card.

In fact the difference would probably be negative since by default, the majority of distros of Linux use the free-software Mesa3D video drivers, not your proprietary (usually faster) closed-source gfx blobs.

>, but the last time i tried compiling z64gl was before purple even created a repo.

I never compiled ziggy's original sources.  ...certainly not on 64-bit Linux, anyway.  I suppose it probably should compile on 32-bit Linux (if that's what you have ... ? I can't remember which Ubuntu I linked you), but definitely purplemarshmallow's fork gave me the Git opportunity to fix the 64-bit portability issues ziggy had, either on Windows or Linux.  So if you had errors...well, gee, did you never see if they were similar ones making Win64 builds?

>Ah I see. Got it compiling now. I was used to just using make. Didn't know I had to do ./ .

You only do `make` if it's a Makefile, not `make.sh`.  make.sh is a Unix shell script, not a Makefile.

It's the same on Windows, except it would be `.\make.sh` with a backslash not forward-.  And that the `.\` is optional on Windows and DOS, because it will last-minute guess the current directory for you if all the system paths don't find it first (which is one of Windows' limitations, despite it sounding convenient).

**Comment by cxd4** [comment]
>that is also assuming hle audio for linux

...What's assuming HLE audio for Linux?  I have no idea what you're referring to.

>but if it was to run on any other platform then it will need more work

Well, it compiles on Linux already, just doesn't run.

So the only way it would need more work to run on any other platforms besides Linux and Windows is if you were using Linux- or Windows-specific code.  You indeed use some Android-specific code for the UI, but I can certainly tell that you are not using Linux-specific code (besides you don't have Linux, right?).

Anything I have done to get Project64 to compile on Linux, was to get Project64 to compile under theoretically any C-conforming operating system at all.  I may not have a Macintosh to test since 16 years ago but I don't care; I don't believe in using Linux-specific code to port Project64 from Windows, just platform-agnostic, standards-conforming practices.

**Comment by LegendOfDragoon** [comment]
>That sounds like you're using the recompiler. I thought I mentioned last year when you were getting it to work on Linux back then, that you cannot use that.

I actually tried m64p 1.5 since I compiled 64 bit. I don't even know / remember how to compile 32bit in gcc with 64bit Linux. I just now remembered that you also told me about 64bit issues with LLE in m64p.. I guess next time I go on Linux, I'll try your mupen fork.

>So if you had errors...well, gee, did you never see if they were similar ones making Win64 builds?

Never tried yet. Haven't used 64 bit Project 64 in a long time. At the time, it didn't occur to me that 64bit might have something to do with the issue.

>It doesn't matter. As far as video plugin bottlenecks go I don't think you should expect a difference when switching between OSes, at any rate. Certainly not on an Intel video card.
In fact the difference would probably be negative since by default, the majority of distros of Linux use the free-software Mesa3D video drivers, not your proprietary (usually faster) closed-source gfx blobs.

I wasn't expecting any miracles but I was hoping at least some stuff runs better. I suppose I'll just have to stick to using faster video plugins.

>It's the same on Windows, except it would be .\make.sh with a backslash not forward-. And that the .\ is optional on Windows and DOS, because it will last-minute guess the current directory for you if all the system paths don't find it first (which is one of Windows' limitations, despite it sounding convenient).

Makes sense now. I'm so used to not having to type in ,\ . On Windows, I usually just use a bat file, which is very convenient imo. Is there a Linux equivalent? I'd prefer double clicking a file over opening terminal and typing commands. I suppose I can write an executable file that sends commands :smile: .

**Comment by cxd4** [comment]
>I actually tried m64p 1.5 since I compiled 64 bit.

You compiled Mupen64 1.5 as 64-bit?

I'm surprised that even launches.  You would have been better off running Hacktarux's pre-compiled binary releases of 32-bit Mupen64 1.5 on 64-bit Ubuntu (with the appropriate 32-bit libraries installed to C:\windows\syswow64 except the Unix-land version of that).

Of course, the 64-bit fork I did with my changes to m64 1.5 should work out of the box, yes.  Either way is fine.  I didn't fix the INI parsing bugs, so you may have to fiddle with the choice of ROM directory to get the ROM browser to remember stuff like that.

>I wasn't expecting any miracles but I was hoping at least some stuff runs better. I suppose I'll just have to stick to using faster video plugins.

On my NVIDIA 8xxx series GeForce laptop, z64gl with LLE graphics and audio ran full speed with most games including the Zelda's and Mario titles.

This newer laptop is some integrated Intel chip...z64gl is even slower than angrylion's plugin on this thing, lol.  I have to use the z64 software rendering for the most speed, and that is slow.  I probably need to disable the free-software Mesa3D drivers on this laptop and use Intel's proprietary ones to get z64gl to run competently on this next laptop I have.

>On Windows, I usually just use a bat file, which is very convenient imo. Is there a Linux equivalent? I'd prefer double clicking a file over opening terminal and typing commands.

The Windows command script to compile my RSP plugin is `make_w32.cmd` IIRC?

The UNIX family shell script is `make.sh`.  They are the same script with slightly different syntax.

**Comment by cxd4** [comment]
You can probably just double-click the make.sh, but I haven't tested this.

The reason it may not necessarily work is you're not formally setting the current directory.

Also you have to Ctrl+Shift+F1 (or Fx, depends) to see the console output from clicking the shell script in a GUI.  Windows pops up a console because it hides it from you; Linux doesn't because by double-clicking the batch script in a GUI desktop you're hiding yourself from the console.

**Comment by cxd4** [comment]
>I'd prefer double clicking a file over opening terminal and typing commands.

I didn't understand when you were saying this.

That isn't modified outside Windows OSes; you can still double-click files to open them.  Did you try checking the output directory (`obj/`) to see that it compiled by double-clicking it?

**Comment by LegendOfDragoon** [comment]
>You compiled Mupen64 1.5 as 64-bit?

I think i downloaded a binary from emutalk.

>I didn't fix the INI parsing bugs, so you may have to fiddle with the choice of ROM directory to get the ROM browser to remember stuff like that.

I guess that explains the issue I had with ROM browser.

>This newer laptop is some integrated Intel chip...z64gl is even slower than angrylion's plugin on this thing, lol. 

Wow that's awful! I at least get decent performance in games like SM64 on Windows with z64gl. I might as well just give up on that plugin then.

>Did you try checking the output directory (obj/) to see that it compiled by double-clicking it?

Couldn't get it to work. It was one of the first things i tried. I tried again today and now it's even associating .sh files as a text file. Not sure what application type to associate it with.

**Comment by cxd4** [comment]
>I think i downloaded a binary from emutalk.

Well then yeah.  The 32-bit binary needs 32-bit plugins.

Back in the day, I had 32-bit Xubuntu and 64-bit Lubuntu.  I would sometimes compile the plugins under 32-bit Ubuntu then copy everything over to the 64-bit partition, and that would work.  Still...if I were you, now that it's been made possible I would tend to just compile everything as 64-bit.

>Wow that's awful! I at least get decent performance in games like SM64 on Windows with z64gl. I might as well just give up on that plugin then.

Nope, same thing with SM64, not even full speed on this laptop.

It **is** full speed if I use the software-rendering z64.so, not z64gl.so, though.  I think it's the same with Windows.  Intel is just horrible with their OpenGL drivers; when I disable Mesa3D on here z64gl will probably go up to full speed but I've just not taken it upon myself to do this yet.

The performance in my OpenGL fork of angrylion's plugin is sometimes better than z64gl on here.  It only does a few GL calls just to blit the texture rectangle, so there's not really any more impact than as if I was on some NVIDIA card.  I did still need to release a new version using texture rectangle optimizations to get it to be faster than the DirectDraw version though.

>I tried again today and now it's even associating .sh files as a text file. Not sure what application type to associate it with.

Ah, well that's a security thing.  I let my distro associate it with text editors, too.  Too much automated execution exposes the immunity of your system.

You might have to right-click the file and pick something like "Execute" or whatever it's called over there.  Make sure the .sh file has the "Execute" permission authorized by you to be allowed in the file properties (right-click Properties, same as Windows).

If Ubuntu doesn't give one, you might have to write your own executable that does `int main() { system("make.sh");` if you really care about being able to double-click it.

Anyway, if you double-click it you would still need to check the terminal for console output in case of any error text.  That's either Ctrl+Alt+F1,2,3,4,5,6 depending on how Ubuntu does things, then it's Ctrl+Alt+F7 to switch back to your X graphical desktop session (which is also, itself, running from the command-line, unlike Windows, which hides the terminal until opening a command-line application, forcing it to shove a console window in front of the screen).

**Comment by cxd4** [comment]
Anyway, it's probably just as easy to right-click the empty background space of the folder, then "Open Terminal Window Here" or whatever it's called, then you're already `cd`'d into the right directory as the `make.sh` file.

And instead of having to actually type out `./make.sh`, just do `./m<TAB>`.  I may type ~94 words per minute, but that doesn't mean relying on tab completion isn't going to end up faster anyway.  Practice it, and you'll be glad you learned it.

>I guess that explains the issue I had with ROM browser.

It happens with some directories.  It's been there since the beginning, but Hacktarux hadn't caught it.

I just put all the ROMs under a root `/n64/` directory and the INI seems to remember that without corrupting the file path with extra characters just fine.  You may have to play with trying different ROM browser paths until you hit one that Mupen64 doesn't insert garbage characters into; I don't know what the pattern is.  As with everything I took the lazy approach and worked around it.

**Comment by cxd4** [comment]
>Wow that's awful! I at least get decent performance in games like SM64 on Windows with z64gl. I might as well just give up on that plugin then.

I've found out that this is caused by tile object creation and deletion.  Intel's MESA driver version seems to suffer most horribly from doing this with texture objects.

Any games with lots of 3-D runs full speed here.  I said SM64 was not running full speed...it turns out this is because of the 2-D texture icons in the HUD.  Nothing 3-D-wise at all seems responsible for the low speed.

I probably just need to orphan z64gl's texture object re-allocation to a different scope to fix it.

**Comment by cxd4** [comment]
And when I said z64gl I meant any plugin at all really, except software renderer's.  Direct3D ones would suffer the same massive speed drop from the same poor driver functions, so it's a little unfair to associate a Linux port of z64gl as being the thing to "give up on" in that event.

**Comment by LegendOfDragoon** [comment]
>Still...if I were you, now that it's been made possible I would tend to just compile everything as 64-bit.

Well i had downloaded 32bit mupen64 0.5 as well as both 32bit and 64 version of m64p 1.5. I saw that the plugins only appeared in the list on the x64 m64p and that's when I realized I've been compiling in x64 this whole time lol. For some reason, I assumed I was always compiling in 32bit lol. I hope to some day move onto 64 bit :smile: .

>The performance in my OpenGL fork of angrylion's plugin is sometimes better than z64gl on here.

Ironically stuff like this is one of the reasons I invested a lot of time into optimizing angrylion's code. I saw that I was getting better performance than hardware renderers in certain scenarios. It almost convinced me that hardware rendering was not worth using.

>Anyway, it's probably just as easy to right-click the empty background space of the folder, then "Open Terminal Window Here" or whatever it's called, then you're already cd'd into the right directory as the make.sh file.
And instead of having to actually type out ./make.sh, just do ./m<TAB>.

Sounds neat. I may just do this. I don't feel like making an exe anymore :smile: . Seems like overkill for something I doubt I'll have to do too often.

>It happens with some directories. It's been there since the beginning, but Hacktarux hadn't caught it.

Ah I see. Maybe I'll try to track down the cause, if I ever decide to start using Linux more often.

>I've found out that this is caused by tile object creation and deletion. Intel's MESA driver version seems to suffer most horribly from doing this with texture objects.

On my desktop, the overall performance with z64gl is much better, but stuff like Copy mode is still not too great. It's slower than z64 for such tasks, iirc.

>And when I said z64gl I meant any plugin at all really, except software renderer's. Direct3D ones would suffer the same massive speed drop from the same poor driver functions, so it's a little unfair to associate a Linux port of z64gl as being the thing to "give up on" in that event.

Have you tried using Rice Video with Intel's MESA driver? On Windows, I get better performance with Rice Video OGL than I do with Jabo's D3D8. I have seen a lot of people blame Windows OS for some OpenGL issues. If the performance is even worse on Linux, then that's the last straw for me. It's bad enough that z64gl has a lot of accuracy issues. If I can't manage to get acceptable performance on my laptop, then I don't have much of a use for it because I can just run more accurate plugins at full speed on my desktop.

If I have to make tons of changes to z64gl's code in order to improve its performance, then I believe I'm better off just writing my own plugin from scratch. I understand that drivers play a big role and seeing Glide64's performance with nGlide has really opened my eyes. However, I still believe part of the problem is the code itself. Even glN64's performance is mostly fine, despite the fact that it's slower than Glide64 on some modern hardware.

**Comment by cxd4** [comment]
>>The performance in my OpenGL fork of angrylion's plugin is sometimes better than z64gl on here.

>Ironically stuff like this is one of the reasons I invested a lot of time into optimizing angrylion's code. I saw that I was getting better performance than hardware renderers in certain scenarios. It almost convinced me that hardware rendering was not worth using.

Yeah but what I said was a little on the extreme end of driver hitches.  z64gl being slower than a software-rendering plugin on this driver is more of z64gl's fault for reallocating tiles in the way it does so much, which could be changed.

A well-tested (for development) hardware renderer should always be faster than the software renderer, but I'm sure it's easy to make a system where it's faster in a plugin to do everything on the CPU and leave the GPU idle instead of using that for any rendering.  Whether we care about such systems is preferential...I think usually we don't.  The software render option is primarily useful just for the reason that this is an emulator for another system's implementation-defined per-pixel RGB colors (e.g. dithering results).

>Sounds neat. I may just do this. I don't feel like making an exe anymore :smile: . Seems like overkill for something I doubt I'll have to do too often.

Yea many people using Ubuntu tend to use it the same ways as they would Windows, very little CLI.

That includes installing software packages with `apt-get` and just going through the Start menu to open the thing every time after that.  You could create a start menu entry for Mupen64 like Windows lets you do.  Same with making a start menu item for invoking the `make.sh`.  I don't care about any of that though; it's pretty easy to use the CLI.

>Have you tried using Rice Video with Intel's MESA driver? On Windows, I get better performance with Rice Video OGL than I do with Jabo's D3D8.

Well the OGL in Rice's plugin is very under-developed and doesn't emulate a lot of effects.  I don't even think it does fog or added triangle shading.

I never built RiceVideo on Linux.  I remember trying to...some unresolved NVIDIA extension function came up in the run-time linker resolution errors.  I wasn't that needy of RiceVideo to track it down and fix it...probably should though but I like LLE more anyway as long as it's still full speed.

>I have seen a lot of people blame Windows OS for some OpenGL issues.

It would be attributable more to the drivers, not the OS.

>If the performance is even worse on Linux, then that's the last straw for me.

Well you're using different drivers though.  Mesa3D is a software-rendering driver with hardware acceleration.

Distribution-wise Linux is really "GNU" in the sense that it normally avoids the proprietary things like the drivers Windows would use.  It's not that they're not available with Linux; it's that everything you get with a fresh install comprises mostly of the open-source stuff.  And the D3D/OGL drivers that are installed on Windows are not open-source; you would have to install them manually with GNU as well because what you're using out of the box on Linux is really Mesa, not the manufacturer's proprietary drivers you tested on Windows.

>If I have to make tons of changes to z64gl's code in order to improve its performance, then I believe I'm better off just writing my own plugin from scratch.

Yeah I haven't found a lot of time in between work days to start that, but I really would rather be doing this.  And seeing it be done by others, not rebranding glN64 or whatever.  None of the things I'd like to enforce in the fundamental design of any OpenGL plugin are in any of the N64 plugins we know.

I still see it as productive and useful to send improvements to existing plugins like z64gl regardless.

>Even glN64's performance is mostly fine, despite the fact that it's slower than Glide64 on some modern hardware.

One of the trickiest things is determining whether or not to use extensions (for modern hardware) or core basics for older hardware.  It often starts to become simplest and cleanest to just remove support for old video hardware...another do-or-don't decision that software rendering solves.  Although, software rendering is never as fast as a well-written hardware-renderer as long as said renderer isn't trying to be pixel-accurate at the same time.

>Ah I see. Maybe I'll try to track down the cause, if I ever decide to start using Linux more often.

Any time you're ready.  And when you do get more comfortable with anything UNIX-like you can say goodbye to a lot of counterproductive users.

**Comment by LegendOfDragoon** [comment]
>A well-tested (for development) hardware renderer should always be faster than the software renderer

Agreed! It really bugs me when others try to say otherwise. They think it's somehow acceptable for intel users to get better performance with a software renderer..

>You could create a start menu entry for Mupen64 like Windows lets you do. 

I gotta find out how to do that.

>Well the OGL in Rice's plugin is very under-developed and doesn't emulate a lot of effects. I don't even think it does fog or added triangle shading.

Yea it's unfortunate that it's under-developed. It does partially support fog although the fog is buggier than the D3D version. Like Jabo 1.6 and glN64, Rice's fog doesn't work at all in certain games like smash bros for some reason :disappointed: . I still think the code is valuable for learning optimizations.

>It's not that they're not available with Linux; it's that everything you get with a fresh install comprises mostly of the open-source stuff.

I wonder if the proprietary driver is any good. I may look into that at some point.

>Yeah I haven't found a lot of time in between work days to start that, but I really would rather be doing this. And seeing it be done by others, not rebranding glN64 or whatever. None of the things I'd like to enforce in the fundamental design of any OpenGL plugin are in any of the N64 plugins we know.
I still see it as productive and useful to send improvements to existing plugins like z64gl regardless.

I currently don't have the time either. I agree that it's productive to send improvements to existing plugins. I just don't know z64gl's code too well, so I'm better off working with other plugins I'm more familiar with.

>And when you do get more comfortable with anything UNIX-like you can say goodbye to a lot of counterproductive users.

Haha that's a nice incentive :smile: .

>It often starts to become simplest and cleanest to just remove support for old video hardware...another do-or-don't decision that software rendering solves

That's another reason why I liked working with a software renderer.

**Comment by project64** [comment]
well I am committed to Glide64 now ... for better or worse .. First trying to get android to be able to set the res, getting close to this in the the glide64 cleanup branch, Then it is angle branch. This will make things worse and introduce bugs that are in the opengl-es version .. but a needed back step. 

After that maybe lle or just trying to fix bugs, but for me I will look at performance after I get lle working nice .. Which is probably studying angry lions plugin since I can not merge it temporarily because of license.

I would like to get the LLE to work in full software like angry lions first.

**Comment by LegendOfDragoon** [comment]
>well I am committed to Glide64 now ... for better or worse ..

I think it's a good investment. Certainly better than GLideN64, for what you are trying to accomplish. I really hope Angle is a good wrapper.

>After that maybe lle or just trying to fix bugs, but for me I will look at performance after I get lle working nice .. Which is probably studying angry lions plugin since I can not merge it temporarily because of license.

I think improving LLE would be good after getting Angle working properly. It should help with debugging and making HLE more accurate. I think having a solid debugger will greatly increase productivity. It seems pretty hard to track down bugs in hardware renderers.

>I would like to get the LLE to work in full software like angry lions first.

What's your incentive for integrating LLE software renderer into Project64 Video Plugin? I'm not sure if it's worth the extra effort. I think I would probably just fork Angrylion's and keep it as a separate project in a different repository.

**Comment by project64** [comment]
one thing I want to do is make the RSP plugin a static lib, be able to link it in to glide64 for testing ... then for a command execute the RSP code for just the command and have it do the matching RSP/RDP so I can interweave hle and lle for testing .. be able to switch at compile time individual commands and see have it work, so doing any hle should work 100%. 

Especially for ucode not handled yet I have the option of having one or two commands lle'd .. 

Partly the reason to go software rendering is that angrylion plugin is the only really accurate good reference for lle. It would be nice to be able to compare software vs hardware renderer for testing/finding bugs.

**Comment by LegendOfDragoon** [comment]
Fair enough, that's more than I assumed you would do with the code. That's an interesting idea. Before you start comparing HLE to LLE, I suggest improving the accuracy of your RSP. Aside from the div instructions, there are a few opcodes that I suspect need to be altered a bit.

I believe that it's only worth doing all that extra work, if you plan to make big changes. I personally don't like to idea of having to use MAME's license when a fair amount of the code was shamelessly copy pasted. If I ever go back to LLE software, I won't bother using that code anymore. I'd rather just read the docs myself and make test ROMs, so I can have full control over my code.

Do you plan on making HLE exactly as accurate as LLE? I think a lot of HLE code has shortcuts to sacrifice accuracy for performance, in addition to human error. I know for audio this is especially true.

>It would be nice to be able to compare software vs hardware renderer for testing/finding bugs.

Yes, it's very helpful. Comparing angrylion's to Glide64 & GLideN64 helped me realize the cause for the clipping issue in Body Harvest. The depth buffer code is simply too inaccurate. It's a prime example of what I mean by the HLE code taking shortcuts, in addition to human error. If the depth buffer code was as accurate as Angrylion's, then it should work. Although instead of integrating code from angrylion into Glide, I currently just debug them separately and compare the values in memory.

**Comment by AmbientMalice** [comment]
>Yes, it's very helpful. Comparing angrylion's to Glide64 & GLideN64 helped me realize the cause for the clipping issue in Body Harvest. The depth buffer code is simply too inaccurate.

Can you elaborate on this? The reason GLideN64 adopted Glide64's software depth buffer was because Gonetz wanted to use code he had ownership over instead of copy-pasting Angrylion's, from memory.

**Comment by LegendOfDragoon** [comment]
>Can you elaborate on this?

Sure. I looked at the data at address 0x802B3800 and compared various plugins and settings.
Here are the end results (in 32bit big endian format) for each plugin after the CPU reads the depth buffer data and modifies it

Glide64 with SW Depth Buffer enabled
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 49 00 3D 00 30 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 4A 00 00 00 00 00 00 00 00 00 00
00 00 00 5C 00 5C 00 5C 00 5B 00 5B 00 4A 00 00
00 00 00 00 00 00 00 B4 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 30 00 3D 00 49
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 41 00 35 00 2C 00 2A 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 4A 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 63 00 63 00 63 00 63 00 4A 00 00
00 00 00 00 00 BC 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 2A 00 2C 00 35 00 41 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

GLideN64 SW Depth Enabled (HLE only)
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 4C 00 54 00 5B 00 63 00 6B 00 73 00 74
00 76 00 77 00 79 00 7B 00 23 00 00 00 00 00 00
00 00 00 00 00 1B 00 20 00 25 00 2A 00 00 00 00
00 00 00 00 00 00 00 28 00 23 00 1E 00 18 00 00
00 00 00 00 00 00 00 AE 00 23 00 7B 00 79 00 77
00 76 00 74 00 73 00 6B 00 63 00 5B 00 54 00 4C
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 52 00 5A 00 62 00 6A 00 70 00 71
00 73 00 75 00 76 00 78 00 79 00 22 00 00 00 00
00 00 00 00 00 1B 00 20 00 25 00 2A 00 2F 00 00
00 00 00 00 00 2D 00 28 00 23 00 1E 00 18 00 00
00 00 00 00 00 00 00 22 00 79 00 78 00 76 00 75
00 73 00 71 00 70 00 6A 00 62 00 5A 00 52 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

Angrylion's
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 4D 00 54 00 5C 00 64 00 6C 00 73 00 74
00 76 00 78 00 79 00 7B 00 7B 00 B0 00 BF 00 BF
00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF
00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF
00 BF 00 BF 00 BF 00 23 00 7A 00 7A 00 79 00 77
00 76 00 74 00 72 00 6A 00 62 00 5A 00 52 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 4F 00 53 00 5B 00 63 00 6A 00 70 00 71
00 73 00 75 00 76 00 78 00 7A 00 7A 00 B5 00 BF
00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF
00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF
00 BF 00 BF 00 7A 00 7A 00 79 00 78 00 76 00 74
00 73 00 71 00 6C 00 68 00 61 00 59 00 51 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00


Here is what the data looks like before it's modified by CPU

GLideN64
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC
FF FC 37 64 33 30 2E FC 2A C4 26 90 22 9C 21 C0
20 E4 20 08 1F 94 1F 24 59 E8 FF FC FF FC FF FC
FF FC FF FC 66 44 5D B0 58 40 52 CC FF FC FF FC
FF FC FF FC FF FC 55 04 5A 74 5F E4 6A B0 FF FC
FF FC FF FC FF FC 11 60 59 E8 1F 24 1F 94 20 08
20 E4 21 C0 22 9C 26 90 2A C4 2E FC 33 30 37 64
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC
FF FC FF FC 33 EC 2F B8 2B 84 27 50 24 2C 23 50
22 70 21 94 20 B8 1F EC 1F 80 5B 6C FF FC FF FC
FF FC FF FC 66 44 5D B0 58 40 52 CC 4D 5C FF FC
FF FC FF FC 4F 94 55 04 5A 74 5F E4 6A B0 FF FC
FF FC FF FC FF FC 5B 6C 1F 80 1F EC 20 B8 21 94
22 70 23 50 24 2C 27 50 2B 84 2F B8 33 F0 FF FC
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC

Angrylion's
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC
FF FC 36 E6 32 BE 2E 92 2A 6A 26 3E 22 8E 21 AE
20 D2 1F FA 1F 8A 1F 32 1F 26 11 0F 0D 08 0D 08
0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08
0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08
0D 08 0D 08 0D 08 5A 1A 1F 4E 1F 36 1F A6 20 2A
21 0A 21 EA 23 26 27 4E 2B 76 2F A2 33 CA FF FC
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC
FF FC 35 86 33 72 2F 46 2B 1E 26 F6 24 1A 23 3A
22 5E 21 7E 20 A2 1F E2 1F 72 1F 66 0F 7B 0D 08
0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08
0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08 0D 08
0D 08 0D 08 1F 36 1F 56 1F 8E 1F FE 20 DA 21 B6
22 96 23 76 25 EE 28 02 2C 2E 30 56 34 7E FF FC
FF FC FF FC FF FC FF FC FF FC FF FC FF FC FF FC

Couldn't be bothered with doing more testing on Glide64. You can see that GLideN64's SW depth buffer code is more accurate than Glide64, but not good enough. 

>The reason GLideN64 adopted Glide64's software depth buffer was because Gonetz wanted to use code he had ownership over instead of copy-pasting Angrylion's, from memory.

That was one of the reasons and I can understand his concern. I think he also liked the fact that his HLE code was faster and simpler. Gonetz wanted the code to work for HLE as well, I think. I tested with GLideN64 Public release 2.0 if version # matters.

**Comment by cxd4** [comment]
>Here are the end results (in 32bit big endian format)

_"Here are the end results (in big-endian format)"_

Fixed that for ya.

>Which is probably studying angry lions plugin since I can not merge it temporarily because of license.

What's the matter with the license?  Last I checked (unless it got redefined again) the MAME license is just a little GPL-compatible free software license, no?  So it should be usable with any GPL-licensed stuff and maybe even re-licensable to some GPL.

>Gonetz wanted to use code he had ownership over instead of copy-pasting Angrylion's, from memory.

That makes little sense.  He had more ownership over angrylion's source than glN64 source, since that was unlicensed and legally default to full copyright protections.  MAME license again is not really that different from a plain GPL...at least to me since none of the restrictions seemed problematic.  I think one of the differences was not being able to sell it for money.

**Comment by gonetz** [comment]
> The reason GLideN64 adopted Glide64's software depth buffer was because Gonetz wanted to use code he had ownership over instead of copy-pasting Angrylion's, from memory.

It never was a reason.

> I think he also liked the fact that his HLE code was faster and simpler. Gonetz wanted the code to work for HLE as well, I think. 

This is true. My depth buffer render works fast, but it usable only for HLE. I tried to adopt it for LLE, but failed.

From the other side, Angrylion's code is only for LLE. GLideN64 LLE mode currently is too slow without software render.
However, if somebody will turn Angrylion's code into a library with simple C interface, it can be used in GLideN64 as optional software render, no problems.

**Comment by AmbientMalice** [comment]
>It never was a reason.

I stand corrected. Sorry.

**Comment by gonetz** [comment]
@LegendOfDragoon 
> Sure. I looked at the data at address 0x802B3800 and compared various plugins and settings.

Could you explain, at which moment CPU reads the depth buffer data and modifies it in Body Harvest?
I tried to run the first mission with Angrylion's plugin and I searched for  0D 08 0D 08 sequence in depth buffer area. It never appeared.

**Comment by LegendOfDragoon** [comment]
@gonetz I apologize for the delayed response.

>I tried to run the first mission with Angrylion's plugin and I searched for 0D 08 0D 08 sequence in depth buffer area. It never appeared.

I think the cpu modifies it shortly after the dlist is done. What you will have to do is set a breakpoint on writing to that area in RDRAM. What I would recommend doing is writing code to that sets a breakpoint when the area being written to is within the address range for that particular area. So in Angrylion's code, I would modify the function fbwrite_16 and would check the value of fb, right before pairwrite.
If you wanted to check the first 4 bytes you would take 0x2B3800 and divide it by 2, and compare that number with fb.

For example, I'd do something like this
`if(fb >= 0x159C00 && fb <= 0x159C02)
 {
//breakpoint code here
}
PAIRWRITE16(fb, rval, hval);`

If you don't mind waiting, within the next few days, I could try debugging with the latest GLideN64 code to see what is going on.

**Comment by gonetz** [comment]
@LegendOfDragoon Thanks for explanation. It helped me to start digging in right direction.

**Comment by LegendOfDragoon** [comment]
You're welcome. I'm glad I was able to help :smile: .
